### DISCUSSION TITLE
chat_history in conversation_retrieval dosen't seem to be working

### DISCUSSION THREAD

**Discussion by MahaleVivek** [ajndkr/lanarky#61] [Q&A] [answered]
when I was trying to implement a solution with conversation_retrieval_chain, I'm getting "A single string input was passed in, but this chain expects multiple inputs ({'question', 'chat_history'}). When a chain expects multiple inputs, please call it by passing in a dictionary, eg `chain({'foo': 1, 'bar': 2})`" However I tried different methods to solve this. but it seems like there error for unsupported chat_history. any help would be appreciated

**Comment by ajndkr** [comment]
Hi! There's already a demo example for conversational retrieval chain. You can find the code here: https://github.com/ajndkr/lanarky/blob/main/examples/app/conversational_retrieval.py

and the usage instructions here: https://github.com/ajndkr/lanarky/tree/main/examples#conversational-retrieval

you can check it out and let me know if it works for you.

**Reply by MahaleVivek** [reply]
thank you for your help! However, I'm facing an issue with memory management in the chain(chat endpoint as well). Even after adding the websocket endpoint, I'm encountering the error 'unsupported chat_history'. Any insights on how to address this issue would be greatly appreciated. Thanks

Based on my understanding, it appears that the current functionality of ConversationRetrievalChain does not include support for WebSockets. Could you please confirm if my understanding is correct?

**Reply by ajndkr** [reply]
> thank you for your help! However, I'm facing an issue with memory management in the chain(chat endpoint as well). Even after adding the websocket endpoint, I'm encountering the error 'unsupported chat_history'. Any insights on how to address this issue would be greatly appreciated. Thanks

hmm. can you please share your full code (or relevant sections) along with the error log? i don't fully understand the issue here.

> Based on my understanding, it appears that the current functionality of ConversationRetrievalChain does not include support for WebSockets. Could you please confirm if my understanding is correct?

yup! `ConversationRetrievalChain` does not have websocket support yet.

**Reply by MahaleVivek** [reply]
![Screenshot from 2023-05-23 16-37-21](https://github.com/ajndkr/lanarky/assets/70280845/df98a252-ec19-4402-b2ca-76a58f3949b7)

The code remains unchanged, just tried to add memory because previous chat memory was not maintained, except for the addition of a Websocket endpoint, which unfortunately is not supported. In the provided example, I observed that when I initially used the input 'Hi,' it returned a response. However, when I followed up with the input 'hello,' it resulted in an error related to an unsupported chat_history format. How can we go about resolving this issue?

**Reply by ajndkr** [reply]
thanks! i will check this out soon. Let me know if you find a solution by time I respond back. Maybe you can contribute a fix (if any).

**Reply by MahaleVivek** [reply]
Thanks, I'll update you if I find a solution.

**Comment by ajndkr** [comment, marked as answer]
@MahaleVivek the `chat_history` field accepts a list of tuples. I've updated the demo example. Please see https://github.com/ajndkr/lanarky/blob/main/examples/README.md#conversational-retrieval

**Comment by lefant** [comment]
I got it to work for websocket by adding a memory like this:
```
        memory = ConversationBufferMemory(
            memory_key="chat_history",
            return_messages=True,
        )
        question_generator = LLMChain(
            llm=ChatOpenAI(
                temperature=0,
                streaming=False, # otherwise synthesised question appears in websocket output...
            ),
            prompt=CONDENSE_QUESTION_PROMPT,
        )
        doc_chain = load_qa_chain(
            llm=ChatOpenAI(
                temperature=0,
                streaming=True,
            ),
            chain_type="stuff",
        )
        return ConversationalRetrievalChain(
            combine_docs_chain=doc_chain,
            question_generator=question_generator,
            memory=memory,
            retriever=db.as_retriever(),
            return_source_documents=False,
            verbose=True,
        )
```

**Reply by ajndkr** [reply]
cool! @lefant would you like to open a PR and contribute?
